### ISSUE TITLE
Stops receiving new messages after a while

### ISSUE THREAD

**Issue by izenja** [xmikos/qhangups#8]
Maybe this should be on the hangups page; let me know if I should post there instead. After connecting, the program works as expected for a while, usually at least 15 minutes but sometimes hours -- I can't be more specific right now. During this time, I usually have conversations open in tabs and they stay open. After a while no new messages are received in these conversations even though the program is connected. If I disconnect and reconnect, the new messages are visible and can be received again.

I have an Android phone with Hangouts connected during this whole time, if that's relevant. I also sometimes sign into Inbox in Firefox, but it doesn't seem like loading Inbox triggers it (I just thought I'd mention it).

I'm using a version pulled from git two days ago (March 30), running Arch Linux. Everything including the needed dependencies were pulled and built on that day through the AUR.

**Comment by izenja** [comment]
I tried using the console hangouts program for a bit, and it eventually exits with a disconnect error. Maybe qhangups isn't catching these and handling it properly? Here's a traceback of the run (of hangups) which encountered the disconnect:

```
Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/hangups/channel.py", line 337, in _longpoll_request
    res.content.read(MAX_READ_BYTES), PUSH_TIMEOUT
  File "/usr/lib/python3.4/asyncio/tasks.py", line 370, in wait_for
    yield from waiter
  File "/usr/lib/python3.4/asyncio/futures.py", line 386, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.4/asyncio/tasks.py", line 287, in _wakeup
    value = future.result()
  File "/usr/lib/python3.4/asyncio/futures.py", line 267, in result
    raise CancelledError
concurrent.futures._base.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/hangups", line 9, in <module>
    load_entry_point('hangups==0.2.6', 'console_scripts', 'hangups')()
  File "/usr/lib/python3.4/site-packages/hangups/ui/__main__.py", line 699, in main
    }, COL_SCHEMES[args.col_scheme])
  File "/usr/lib/python3.4/site-packages/hangups/ui/__main__.py", line 76, in __init__
    loop.run_until_complete(self._client.connect())
  File "/usr/lib/python3.4/asyncio/base_events.py", line 316, in run_until_complete
    return future.result()
  File "/usr/lib/python3.4/asyncio/futures.py", line 275, in result
    raise self._exception
  File "/usr/lib/python3.4/asyncio/tasks.py", line 234, in _step
    result = coro.throw(exc)
  File "/usr/lib/python3.4/site-packages/hangups/client.py", line 121, in connect
    yield from self._listen_future
  File "/usr/lib/python3.4/asyncio/futures.py", line 386, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.4/asyncio/tasks.py", line 287, in _wakeup
    value = future.result()
  File "/usr/lib/python3.4/asyncio/futures.py", line 275, in result
    raise self._exception
  File "/usr/lib/python3.4/asyncio/tasks.py", line 234, in _step
    result = coro.throw(exc)
  File "/usr/lib/python3.4/site-packages/hangups/channel.py", line 204, in listen
    yield from self._longpoll_request()
  File "/usr/lib/python3.4/site-packages/hangups/channel.py", line 341, in _longpoll_request
    except aiohttp.errors.ConnectionError as e:
AttributeError: 'module' object has no attribute 'ConnectionError'
Exception ignored in: %    
```

**Comment by tdryer** [comment]
Can you check that you have the correct version of `aiohttp` installed? hangups requires 0.9.1.

**Comment by izenja** [comment]
I have 0.14.4. I'll downgrade to 0.9.1 and see what happens.

**Comment by izenja** [comment]
That didn't help. I should have mentioned before, I can still send messages through the conversation window after the issue happens. They get received by the others, I just get no new messages on my end.

I just noticed it seemed to happen when someone sent emoticons. I did some testing and that seems to be the case. When one of the emoji is sent by someone through the hangouts Android app, it triggers the issue. It could be some sort of UTF issue. It doesn't seem to affect the console version, and I think the disconnects that I mentioned earlier are not relevant. Qhangups does tell me when a disconnect and reconnect occur.

There's nothing on console output or in the log file when it breaks.

**Comment by tdryer** [comment]
Any emoticons in particular? They should work fine with hangups. Do you get a different stacktrace with aiohttp 0.9.1?

**Comment by izenja** [comment]
I'm not sure about the standard ones like :) but the more "creative" ones like guitars and such all seemed to do it. The stack trace only happened in the console version and I haven't gotten a disconnect yet so I don't know if it's any different.

I messed around with my system locale settings (I have /etc/locale.conf set to use UTF-8 but KDE Plasma5 seems to be removing the UTF-8 part) and the emoticons aren't triggering it anymore. I suspect that locale issues are at fault here, but I'll need some more time to see if I can fix the locale and if qhangups breaks after prolonged use after that.

**Comment by xmikos** [comment]
That's definitely the problem, I have already noticed in the past that Hangups only works reliably if you have UTF-8 in locales. But I also use KDE 5 (in Arch Linux) and it isn't messing with my locales. It respects locales set by my system.

**Comment by izenja** [comment]
I think that was the problem - things have been working fine since then.

I did have locale set up seemingly correctly to en_US.UTF-8 based on the Arch wiki, but KDE was dropping the UTF-8 part and overriding the system setting for whatever reason. What I did is delete plasma-localerc and plasma-locale-settings.sh, remove the non-UTF8 line from locale.gen, enable en_CA.UTF-8 in the .gen as well (and run locale-gen), and set KDE to use Canadian English (since I'm Canadian anyway). I'm not sure what did the trick but the KDE locale script is now setting UTF8 and Hangups is working properly.
